### PR TITLE
Support for CTI Astro Rev G TX2 Carrier Board

### PR DESCRIPTION
**Pull request by keenanjohnson**
This adds the DTB and build references to enable builds for CTI Astro revision G boards.

I took the files from https://github.com/balena-os/balena-jetson-skx2/pull/200 and created a fresh branch with a stylistically correct commit message in hopes of finishing the work that @c-wilson started.

Let me know if this all looks correct and/or if I need to update anything else? I'm new to the project, so still figuring out all the nuances here.

Thanks! I'm really excited to help out here!

### CODE REVIEW

**Comment by acostach** [comment]
Thanks for moving the PR here @keenanjohnson ,

A couple suggestions would be:
- let's use the CTI Astro TX2 name. 'Rev G' will need to be removed from all the places it was used
- use yocto-warrior in the coffee file https://github.com/balena-os/balena-jetson/pull/71/commits/d757cc200ffef5c9c882b2164a04f2774bbc5c63#diff-a94f96cc6de2cc137ea79f17709fc6bcR44
- Add a "Changelog-entry" to just one of your commits, so automatic versioning of the PR can be performed. No need to open a new PR for these changes, you can use git rebase --interactive to alter the commits and then force push the branch.

To test your local build you can use 'balena config inject ' to add your TX2 dashboard App config.json to the flasher sd-card: https://www.balena.io/docs/reference/balena-cli/#config-inject-file

I think CTI offers some 32.3.1 dtbs for the Astro carrier board, you can check see if that version works with this PR. Before flashing the board with the PR, please first use the 32.3.1 or 32.4.2 BSP archive to flash the DTBs to the board, so that the kernel on your sd-card can boot with the dtbs from the emmc.

**Comment by keenanjohnson** [comment]
Thanks I’ll make those changes! I understand how to inject the config but what device type should I use to create the config in my balena dashboard given that I can’t select my newly created device type there?

**Comment by keenanjohnson** [comment]
@acostach I've changed the yocto version, revised the name to be just astro-tx2, and added the changelog entry. Let me know how that looks!

**Comment by acostach** [comment]
@resin-jenkins test this please

@keenanjohnson You can choose the Jetson TX2 type to create the App and then you can use the config.json from it.

**Comment by keenanjohnson** [comment]
Is the balena config inject different than the balena preload command?

I attempted to flash the image I built above after using the balena preload command with my app name. However, when the board booted, it booted in Gnome instead of Balena. 

Any ideas on what might have gone wrong or methods to troubleshoot this?

Thank you!

**Comment by acostach** [comment]
Hi @keenanjohnson,

'balena config inject' adds the config.json that you specify to the image you will use for flashing. 'balena preload' downloads your APP image and embeds it into the flasher sd-card, so that when the board comes online, it already has the container APP and doesn't need to download it. 

The fact that the board booted to gnome means that the balenaOS image was probably not flashed to the device when you booted with the sd-card inserted. One reason that this might have happened could be that you were using a different L4T - did you flash l4t 32.4.2 on the target before booting it with the sd-card inserted? Please see my first comment above for details on this. 

You can also boot with a serial console attached to see if the board boots using your flasher sd-card.

**Comment by keenanjohnson** [comment]
Oh no I guess I had started from the stock nividia jetpack. What does the CTI image do that the Nvidia one does not?

Also, what's the easiest command to mount the Linux image?

**Comment by keenanjohnson** [comment]
When I try running the local config option I get the an unsupported file system error.  It looks like there is a support thread mentioning this is not supported for Jetson Nano image.

https://forums.balena.io/t/deploy-locally-built-image/35914/17

`Keenans-MBP:Downloads keenan$ sudo balena config inject testbed-tx2.config.json --type jetson-tx2 --drive /dev/disk2
Password:
Unsupported filesystem.

Additional information may be available by setting a DEBUG=1 environment
variable: "set DEBUG=1" on a Windows command prompt, "$env:DEBUG = 1" on
powershell, or "export DEBUG=1" on Linux or macOS.

If you need help, don't hesitate in contacting our support forums at
https://forums.balena.io

For CLI bug reports or feature requests, have a look at the GitHub issues or
create a new one at: https://github.com/balena-io/balena-cli/issues/

Keenans-MBP:Downloads keenan$ 
`

**Comment by keenanjohnson** [comment]
Ah I found the answer here: https://forums.balena.io/t/balena-config-inject-failing-for-local-tx2-built/123631/10

**Comment by keenanjohnson** [comment]
Ok I got the image to boot and do the provisions step. However, after removing the SD card, the image doesn't boot and I see a kernel panic.

I've attached the log here. Any suggestions for how to debug?

[balena bootlog sdcard copies.txt](https://github.com/balena-os/balena-jetson/files/4714170/balena.bootlog.sdcard.copies.txt)

**Comment by acostach** [comment]
@keenanjohnson From the log you pasted it looks like the DTB you are using might not be compatible with L4T 32.4.2, I saw the exact same error when I was attempting a L4T 28.3 DTB with a L4T 32.2 kernel:

```
[    0.000000] Linux version 4.9.140-l4t-r32.4.2 (oe-user@oe-host) (gcc version 8.3.0 (GCC) ) #1 SMP PREEMPT Sun May 31 00:59:55 UTC 2020
[    0.780531] iommu_context_dev 13e10000.host1x:ctx0: iommu is not enabled for context device. aborting.
[    0.790296] iommu_context_dev 13e10000.host1x:ctx1: iommu is not enabled for context device. aborting.
[    0.800039] iommu_context_dev 13e10000.host1x:ctx2: iommu is not enabled for context device. aborting.
```

When you first provisioned the board using the sd-card, the kernel for 32.4.2 was loaded from sd-card, the DTB was also 32.4.2 but on the emmc and was compatible with the kernel, that's why it booted and provisioned. However, after removing the sd-card and rebooting, the DTB was updated on the emmc with the one you used to build the image.

**Comment by keenanjohnson** [comment]
Gotcha. So the situation is that CTI doesn't have a board support package for 32.4.2 yet, so I don't have an officially compatible device tree. 

How is this done for the other CTI boards right now? I know there isn't a 32.4.2 device tree from CTI that supports the Spacely board for instance, so how is it supported in Balena? https://github.com/balena-os/balena-jetson/blob/master/spacely-tx2.coffee

File: layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb [hunk 1]
```diff
@@ -222,6 +223,7 @@ do_compile[depends] += " \
     virtual/bootloader:do_deploy \
 "
 
+do_compile[depends] += "u-boot-tegra:do_deploy"
```
Comment (keenanjohnson) [inline layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb:226]: Do I need this uboot line? This was left over from the original PR someone added. 

Can you help me understand what this does @acostach ? Thanks!

Comment (acostach) [reply layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb:226]: sure, you notice it's not do_compile_<machine_name>[depends] but do_compile[depends], so it's meant to apply to every device type that uses this package. As the Astro yocto machine extends the jetson-tx2, then this line applies to it too.

What it does actually is to make sure that u-boot gets built and deployed before this tegra186-flash-dry recipe can use the compiled u-boot binary to sign it. On TX2, DTBs and u-boot are signed by nvidia host binaries and then written to raw partitions on the emmc. Signing in this case is performed by
https://github.com/balena-os/balena-jetson/blob/c2d2576bcad333fe3815186ef3d26ba123ef68d5/layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb#L130
which gets in  command line
https://github.com/balena-os/balena-jetson/blob/c2d2576bcad333fe3815186ef3d26ba123ef68d5/layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb#L136
Therefore it parses the xml, finds u-boot there and performs the actual signing on it:
https://github.com/balena-os/balena-jetson/blob/c2d2576bcad333fe3815186ef3d26ba123ef68d5/layers/meta-balena-jetson/recipes-bsp/tegra-binaries/files/resinOS-flash186.xml#L362
and after the u-boot binary is signed, it becomes:
https://github.com/balena-os/balena-jetson/blob/c2d2576bcad333fe3815186ef3d26ba123ef68d5/layers/meta-balena-jetson/recipes-bsp/tegra-binaries/files/partition_specification186.txt#L20

Comment (keenanjohnson) [reply layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb:226]: Got it thanks!

Comment (keenanjohnson) [reply layers/meta-balena-jetson/recipes-bsp/tegra-binaries/tegra186-flash-dry_32.4.2.bb:226]: Do I need it in this review? Obviously the other boards are working fine without it?

**Comment by keenanjohnson** [comment]
Ok so looking through the official board support packages, I realized that the dtb files which control the USB and ethernet interfaces are common to the the Spacely carrier board. So just for kicks, I downloaded the image for the CTI spacely board and was able to load that onto my astro board stack and everything worked perfectly: both ethernet and usb interfaces were functioning.

**Comment by keenanjohnson** [comment]
Update: Digging through the CTI BSP drops, I noticed that the device trees for the ethernet and USB interfaces are common for the CTI Spacely board and the Astro board. Given that the Spacely board is already supported by Balena, I decided to try booting the Spacely image on my Astro stack. This worked perfectly. Everything booted and I confirmed that both Ethernet and USB interfaces were functional.

Given that, I'm going to update my PR here to reference the device tree that's already committed for the Astro board. This is confirmed to be working and ready for review.

**Comment by floion** [comment]
My advice would be to use the dtb that the vendor created for the board. You may be fine now with the dtb from another board but you may soon notice some things not working as expected because you are using dtbs from another board.

**Comment by keenanjohnson** [comment]
Yes I agree, but I still am confused as to why the Spacely board is supported given that CTI doesn't have an official BSP for any of their boards on 32.4.2 yet? Is Balena still supporting older NVIDIA releases somehow or is this just luck? I'm trying to figure out if I need to wait for CTI to release the BSP or if there is any way to make progress before then?

**Comment by acostach** [comment]
@keenanjohnson Can you please try do a clean local build of https://github.com/balena-os/balena-jetson/pull/74 ? 

I've added there the astro machine you PRed, and included the 32.3.1 Astro Rev G+ dtb - the latest available. Please let me know if it works and unblocks things on your side until CTI releases newer 32.4.2 dtbs

**Comment by keenanjohnson** [comment]
Yes I can try that!

**Comment by keenanjohnson** [comment]
Question, once a valid balena distribution is installed, to install a new local build do I need to reinstall Jetpack or are the needed scripts to copy the SD card image included in the Balena base image?

**Comment by keenanjohnson** [comment]
Ok I was finally able to complete a local build today to test. Sorry for the delay but the builds take about 8 hours on my machine here right now and it kept running out of disk space. 

Unfortunately, it looks like the jetpack version didn't copy the img from the SD card when I insert it. I'll debug that further today.

**Comment by acostach** [comment]
@keenanjohnson not sure you already use sstate and shared downloads. With them you can cut down build times significantly. For instance, you can start a build like so:

`balena-yocto-scripts/build/barys -b my_build_dir -m jetson-tx2 --shared-sstate /home/user/work/sstate-cache/ --shared-downloads /home/user/work/downloads/ -d`

Afterwards, if you want to rebuild after pulling some commits, or you checkout a branch, you can do 
`git submodule update --init --recursive` 

and then to re-build the image:

`cd my_build_dir`
`source ../layers/poky/oe-init-build-env .`
`MACHINE=jetson-tx2 bitbake resin-image-flasher -c cleansstate && MACHINE=jetson-tx2 bitbake resin-image-flasher`

The resulting image will be available inside the build directory, in my_build_dir/tmp/deploy/images/jetson-tx2/resin-image-flasher-jetson-tx2-\<timestamp\>.rootfs.resinos-img

**Comment by acostach** [comment]
> Question, once a valid balena distribution is installed, to install a new local build do I need to reinstall Jetpack or are the needed scripts to copy the SD card image included in the Balena base image?

If the valid distribution that is running on the board is at the same l4t as your local build, it shouldn't be necessary as long as the local can run from the sd-card to flash the board. The flasher image contains all the necessary scripts to flash the emmc. You can check the current l4t version on your board with uname -r and in your local build it's in the tegra186-flash-dry.bb recipe name.

**Comment by keenanjohnson** [comment]
Ok so I've built this image and tried to get it to work, but for some reason the Nvidia scripts won't copy this image into the onboard memory. I've confirmed that I can reimage the sd card with the spacely image and that works fine but not this local build. 

CTI released their BSP for 32.4.2 today, so I'm going to try this build with that device tree next.

**Comment by keenanjohnson** [comment]
@acostach that is super helpful about the shared downloads, no I hadn't been doing that. 

I'm not sure why the flasher scripts didn't pull my local image above though. I can understand it not booting after it was copied, but not even being flashed?

**Comment by acostach** [comment]
> CTI released their BSP for 32.4.2 today, so I'm going to try this build with that device tree next.

@keenanjohnson for TX2 carrier boards like Astro, Spacely, Orbitty I see dtbs just for 32.3.1.
Only for the NX there's a 32.4.2 BSP available on the website.

**Comment by acostach** [comment]
Also @keenanjohnson I've succesfully tested a #74 astro-tx2 local build on a TX2 devkit. The board was flashed succesfully and it booted. I can see the astro dtb is present on the emmc:
```
root@165069f:~# strings /dev/disk/by-partlabel/kernel-dtb | grep cti  
ategra186-tx2-cti-ASG001-revG+.dts
```
which makes me believe it should work for you too. But first, please ensure you download https://developer.nvidia.com/embedded/L4T/r32_Release_v4.2/t186ref_release_aarch64/Tegra186_Linux_R32.4.2_aarch64.tbz2 , unpack it, cd Linux_for_Tegra/ && sudo flash.sh jetson-tx2 mmcblk0p1. 

After the board gets flashed, insert your the card on which you wrote your resin-image-**flasher**-...resinos.img and reboot the board so it can start flashing. If you have a serial cable, during flashing you can check the status with systemctl status resin-init-flasher .

**Comment by keenanjohnson** [comment]
Sure I'll try again today.

**Comment by keenanjohnson** [comment]
Is it possible for you to send me the same built image you tested with?

**Comment by keenanjohnson** [comment]
Actually nevermind. I've completed my testing with the build from #74 that works perfectly :).

I've confirmed both USB ports, both ethernet ports, and GPIOs, so let's push #74 through :)

**Comment by keenanjohnson** [comment]
I also diff'd the new .dtb file from the latest CTI BSP and it's the same as the one you've included in #74. What's the best way to proceed for integrating this into the mainline balena support?

**Comment by acostach** [comment]
@keenanjohnson Thanks for letting me know, I've updated the dtb for Astro to the new CTI 32.4.2 release. After the PR is reviewed and merged we'll add the board to the staging dashboard.

**Comment by keenanjohnson** [comment]
Great! Thanks for all your help in this process! I couldn’t have done it
without you.

Let me know if I should close this PR and/or how I can help validate things
further after your PR is reviewed.

**Comment by keenanjohnson** [comment]
Implemented by #74 :)

**Comment by acostach** [comment]
@keenanjohnson wanted to let you know that the astro-tx2 is now available in the staging dashboard

**Comment by keenanjohnson** [comment]
That's great news! I see it and I'm loading it up to test now :)

**Comment by keenanjohnson** [comment]
FYI I tested the image on staging and everything looks great!

**Pull request closed without merging**